figma/lib/fetch_components.ts

```diff
@@ -0,0 +1,246 @@
+import { Node } from 'figma-api'
+import Figma from './client'
+import AVAILABLE_THEMES from '../../src/lib/shared/themes'
+const THEME_KEYS = AVAILABLE_THEMES.map(({ key }) => key)
+import { getWalk, getFileTopLevelChildren, camelize, ExtractorFn, pipe, objToArr } from './helpers'
+
+const BOOL_LABELS = ['true', 'false', 'on', 'off', 'yes', 'no']
+const DEFAULT_LABEL = 'default'
+const TEXT_LIKE_PROP: PropData = { required: false, type: ['string', 'JSX.Element', 'null'], name: 'label' }
+const KNOWN_PROP_TYPES: Props = {
+  label: TEXT_LIKE_PROP,
+  content: TEXT_LIKE_PROP,
+  text: TEXT_LIKE_PROP,
+  children: TEXT_LIKE_PROP,
+  hint: TEXT_LIKE_PROP,
+  placeholder: TEXT_LIKE_PROP,
+}
+function stripBooleanValues(values: string[]): string[] {
+  return values.filter((value) => !BOOL_LABELS.includes(value.toLowerCase()))
+}
+
+const getComponent: ExtractorFn<Node<'COMPONENT'>, 'COMPONENT'> = (node: Node) => {
+  if (node.type !== 'COMPONENT') return
+  return [node.id, 'COMPONENT', node as Node<'COMPONENT'>] as const
+}
+
+const getFrame: ExtractorFn<Node<'FRAME'>, 'FRAME'> = (node: Node) => {
+  if (node.type !== 'FRAME') return
+  return [node.id, 'FRAME', node as Node<'FRAME'>] as const
+}
+
+const getComponentSet: ExtractorFn<Node<'COMPONENT_SET'>, 'COMPONENT_SET'> = (node: Node) => {
+  if (node.type !== 'COMPONENT_SET') return
+  return [node.id, 'COMPONENT_SET', node as Node<'COMPONENT_SET'>] as const
+}
+
+const extractors = [getComponent, getFrame, getComponentSet] as const
+
+export function getComponentSpecUrl(
+  { name: name_, id: fileId }: FileInfo,
+  { id: id_, themes }: ComponentInfo,
+  theme?: string
+) {
+  const name = name_.replace(/\s/gi, '-')
+  const id = (theme && themes[theme]?.id) ?? id_
+  return `${process.env.STORYBOOK_FIGMA_URL}/file/${fileId}/${name}?node-id=${encodeURIComponent(id)}`
+}
+
+export type PropType = 'string' | 'boolean' | 'number' | 'object' | 'array' | 'null' | 'JSX.Element'
+export type PropTypes = Array<PropType>
+export type PropData = { required: boolean; name: string; values?: string[]; type?: PropTypes; items?: PropData }
+export type Props = Record<string, PropData>
+type RawComponentsType = {
+  variants: Record<string, Pick<Node<'COMPONENT'>, 'name'>>
+  theme?: string
+}
+export type ThemeData = Record<string, { id: string; theme: string }>
+type MergedComponentType = Omit<RawComponentsType, 'theme'> & { themes?: ThemeData }
+
+type ComponentsType = { states: string[]; props: Props; themes?: ThemeData }
+type CommonComponentInfo = Pick<Node<'COMPONENT_SET'>, 'name' | 'id'> & {
+  path: string
+  canvasName: string
+  frameId: string
+  frameName: string
+}
+type RawOutput<Type extends {} = RawComponentsType> = Record<string, RawComponentInfo<Type>>
+type RawComponentInfo<T extends {}> = CommonComponentInfo & T
+
+export type FileInfo = { name: string; id: string }
+
+export type Output = RawOutput<ComponentsType>
+export type ComponentInfo = RawComponentInfo<ComponentsType>
```

Comment:
Yeah, part of the cleanup would be to move it to `types.ts`, but honestly I want to a lot of that code to the generator folder and think about the whole organisation, because it's becoming a bit unwieldy. So yes in principle.